core/scripts/chaincli/handler/keeper_upkeep_history.go

```diff
@@ -0,0 +1,173 @@
+package handler
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"log"
+	"math/big"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/smartcontractkit/chainlink/core/services/keeper"
+	"github.com/smartcontractkit/chainlink/core/services/keystore/keys/ethkey"
+)
+
+const (
+	defaultMaxBlocksRange = 1000
+)
+
+// UpkeepHistory prints the checkUpkeep status and keeper responsibility for a given upkeep in a set block range
+func (k *Keeper) UpkeepHistory(ctx context.Context, upkeepId int64, from, to uint64) {
+	// There must not be a large different between boundaries
+	if to-from > defaultMaxBlocksRange {
+		log.Fatal("blocks range difference must not more than 1000")
+	}
+
+	registryAddr, registryClient := k.GetRegistry(ctx)
+
+	// Get positioning constant of the current registry
+	positioningConstant, err := keeper.CalcPositioningConstant(upkeepId, ethkey.EIP55AddressFromAddress(registryAddr))
+	if err != nil {
+		log.Fatal("failed to get positioning constant: ", err)
+	}
+	log.Println("Calculated Positioning Constant for the registry: ", positioningConstant)
+
+	callOpts := bind.CallOpts{Context: ctx}
+
+	log.Println("Fetching registry config")
+	registryConfig, err := registryClient.GetConfig(&callOpts)
+	if err != nil {
+		log.Fatal("failed to fetch registry config: ", err)
+	}
+	blockCountPerTurn := registryConfig.BlockCountPerTurn.Uint64()
+
+	log.Println("Fetching registry keepers list")
+	keepersList, err := registryClient.GetKeeperList(&callOpts)
+	if err != nil {
+		log.Fatal("failed to fetch keepers list: ", err)
+	}
+
+	log.Println("Preparing a batch call request")
+	var reqs []rpc.BatchElem
+	var results []*string
+	var keeperPerBlock []uint64
+	for block := from; block <= to; block++ {
+		keeperIndex := (uint64(positioningConstant) + ((block - (block % blockCountPerTurn)) / blockCountPerTurn)) % uint64(len(keepersList))
+		payload, err := keeper.RegistryABI.Pack("checkUpkeep", big.NewInt(upkeepId), keepersList[keeperIndex])
```

Comment:
We might need to rethink this a little bit... I'm concerned about the case where `checkUpkeep` reverts with `keepers must take turns` _but_ `checkUpkeep` returns `true` from the customer contract. I think as is, the CLI will show `checkUpkeep = false` and revert reason `keepers must take turns` .... but really I _think_ we will want to show `checkUpkeek = true` and the same revert reason.